src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedEntryPointSymbol.cs

```diff
@@ -329,6 +335,153 @@ private static BoundCall CreateParameterlessCall(CSharpSyntaxNode syntax, BoundE
+                    var stringType = compilation.GetSpecialType(SpecialType.System_String);
+                    var stringArrayType = ArrayTypeSymbol.CreateCSharpArray(compilation.Assembly, stringType);
+                    _parameters = ImmutableArray.Create(
+                        SynthesizedParameterSymbol.Create(this, stringArrayType, 0, RefKind.None, userMain.Parameters[0].Name));
+                }
+                else
+                {
+                    Debug.Assert(userMain.ParameterCount == 0);
+                    _parameters = ImmutableArray<ParameterSymbol>.Empty;
+                }
+
+                var userMainLocation = userMain.DeclaringSyntaxReferences.SingleOrDefault()?.GetLocation() ?? NoLocation.Singleton;
+
+                _getAwaiterMethod = GetRequiredMethod(_userMain.ReturnType, WellKnownMemberNames.GetAwaiter, diagnosticBag, userMainLocation);
+                if ((object)_getAwaiterMethod != null)
+                {
+                    _getResultMethod = GetRequiredMethod(_getAwaiterMethod.ReturnType, WellKnownMemberNames.GetResult, diagnosticBag);
+                }
+            }
+
+            public override string Name => MainName;
+
+            public override ImmutableArray<ParameterSymbol> Parameters => _parameters;
+
+            internal override BoundBlock CreateBody()
+            {
+                var syntax = this.GetSyntax();
+
+                var arguments = Parameters.SelectAsArray(p => (BoundExpression) new BoundParameter(syntax, p, p.Type));
+
+                // Main(args) or Main()
+                BoundCall userMainInvocation = new BoundCall(
+                        syntax: this.GetSyntax(),
+                        receiverOpt: null,
+                        method: _userMain,
+                        arguments: arguments,
+                        argumentNamesOpt: default(ImmutableArray<string>),
+                        argumentRefKindsOpt: default(ImmutableArray<RefKind>),
+                        isDelegateCall: false,
+                        expanded: false,
+                        invokedAsExtensionMethod: false,
+                        argsToParamsOpt: default(ImmutableArray<int>),
+                        resultKind: LookupResultKind.Viable,
+                        type: _userMain.ReturnType)
+                { WasCompilerGenerated = true };
+
+
+                // GetAwaiter().GetResult()
+                BoundCall getAwaiterGetResult =
+                    CreateParameterlessCall(
+                            syntax: syntax,
+                            method: _getResultMethod,
+                            receiver: CreateParameterlessCall(
+                                syntax: syntax,
+                                method: _getAwaiterMethod,
+                                receiver: userMainInvocation
+                            )
+                    );
+
+                BoundStatement statement = null;
+                switch (_returnKind)
```

Comment:
Is `_returnKind` necessary or could this be replaced with `if (ReturnsVoid) { ... } else { ... }`?